students/C_Farris/session09/OO_Mailroom.py

```diff
@@ -0,0 +1,194 @@
+                self.donor_dict[donor_name].add_donation(newDonation)
+
+
+    def create_report(self):
+        """returns data for report"""
+        Header = ['Donor Name', 'Total Donation', 'Number Donations', 'Average Donation']
+        Header_string = f'\n{Header[0]:20} {Header[1]:<15} {Header[2]:>5} {Header[3]:<25}'
+
+        donorReport = [[round(float(sum(value.donations)), 2), key, len(value.donations),
+                    round(float(sum(value.donations) / len(value.donations)), 2)]
+                    for key, value in self.donor_dict.items()]
+        sortedReport = sorted(donorReport)
+        ascendingReport = sortedReport[::-1]
+        return Header, Header_string, ascendingReport
+
+    def prepare_to_write_to_disk(self, out_path=OUT_PATH):
+        """Check OUT_PATH specified is a directory, if not, it will make it
+        This is to run once at the start of the application.
+        :Param: none
+        :Return: none. Creates specified folder in CWD if not already created.
+        """
+        if not os.path.isdir(out_path):
+            os.mkdir(out_path)
+
+    def send_file_to_disk(self, key, donorLetter):
+        """
+        Accepts "Thank You" donor Letter and writes to disk.
+        :param (person): Name of person, will be used in file name
+        :param (donorLetter): Thank you letter to donor
+        :return: none. Will send file to directory set in outpath.
+        """
+        filename = key.replace(' ', '_') + '.txt'
+        filename = os.path.join(OUT_PATH, filename)
+        open(filename, 'w').write(donorLetter)
+
+
+    def save_all_thank_yous(self, out_path = OUT_PATH):
+        """returns thank you's for all donors in donor collection
+           creates files with donor_name.txt and saves them in a directory 
+           titled by default "thank_you_letters."""
+
+        for key, value in self.donor_dict.items():
+            self.prepare_to_write_to_disk(out_path)
+            self.send_file_to_disk(key, value.thank_your_letter())
+
+
+
+class Donor ():
+
+    def __init__(self, name, donations=None): ##change donations to none, 
+        self.name = name
+
+        if donations is None:
+            self.donations = []
+        else:
+            try:
+                self.donations = list(donations)
+            except TypeError:
+                self.donations = [donations]
+
+    def __repr__(self):
+        return f'{self.__class__.__name__}({self.name}) : {self.donations}'
+
+    def add_donation(self, donation):
+        """Adds a new donation to the donor list. 
+        Validation requires amount to be a penny or greater and 
+        a numeric value.""" 
+        try:
+            value = donation/2
+        except TypeError:
+            return "Donation value must be entirely numeric!"
+        else:
+            if value <=0.005:
+                return "Donation cannot be smaller than a penny!"
+            else:
+                self.donations.append(donation)
```

Comment:
Something odd going on here.  When adding a donation, it's overwriting the existing donation, rather than adding to it.